fairseq/optim/multiobj/multiobj_sgd.py

```diff
@@ -0,0 +1,277 @@
+import torch as th
+from torch.optim.optimizer import Optimizer, required
+from . import register_multiobj_optim
+
+
+def normalize_param(W):
+    return W / W.norm(2).clamp(min=1e-12)
+
+
+def to_vector(tensors):
+    """Flatten a list of parameters/gradients to a vector"""
+    return th.cat([t.view(-1) for t in tensors])
+
+
+def from_vector(tensors, vector):
+    """Reverse `to_vector` (overwrites the tensor values)"""
+    pointer = 0
+    for tensor in tensors:
+        new_val = vector[pointer:pointer+tensor.numel()].view(tensor.size())
+        tensor.copy_(new_val)
+        pointer += tensor.numel()
+
+
+class MultiObjSGD(Optimizer):
+    """
+    This optimizer works like SGD excepts:
+
+    1. it stores gradient from an auxiliary task with `.save_auxiliary()`
+    2. it uses those auxiliary gradients using `.combine_gradientss()` before
+        applying the update
+
+    Args:
+        full_gradients (bool): do gradient combination ops on the full
+            gradients (as opposed to separately for each parameter)
+    """
+
+    def __init__(self, params, lr=required, momentum=0, dampening=0,
+                 weight_decay=0, nesterov=False, always_project=True,
+                 full_gradients=False):
+        if lr is not required and lr < 0.0:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if momentum < 0.0:
+            raise ValueError("Invalid momentum value: {}".format(momentum))
+        if weight_decay < 0.0:
+            raise ValueError(
+                "Invalid weight_decay value: {}".format(weight_decay))
+
+        defaults = dict(lr=lr, momentum=momentum, dampening=dampening,
+                        weight_decay=weight_decay, nesterov=nesterov,
+                        frozen=False)
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError(
+                "Nesterov momentum requires a momentum and zero dampening")
+        super(MultiObjSGD, self).__init__(params, defaults)
+        self.always_project = always_project
+        self.full_gradients = full_gradients
+
+    def __setstate__(self, state):
+        super(MultiObjSGD, self).__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("nesterov", False)
+
+    def save_auxiliary(self):
+        """This saves the gradients wrt. the auxiliary objective"""
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                param_state = self.state[p]
+                # skip frozen parameters (TODO: remove this)
+                if getattr(param_state, "frozen", False):
+                    continue
+                # Actually save the gradient
+                param_state["aux_grad"] = th.zeros_like(p.data)
+                if p.grad is None:
+                    continue
+                d_p = p.grad.data
+                param_state["aux_grad"].add_(d_p)
```

Comment:
Is this is the case that there's alway one auxiliary gradient?

Reply:
For now yes. I am planning to add support for multiple auxiliary gradients later but this will complicate the code so I'm leaving it for later.